paasta_tools/setup_prometheus_adapter_config.py

```diff
@@ -198,75 +196,71 @@ def _minify_promql(query: str) -> str:
-
-        return True, None
+def _uses_prometheus(provider: Optional[MetricsProviderDict]) -> bool:
+    if provider is None:
+        return False
 
-    return False, "did not request uwsgi autoscaling"
+    # Check provider-specific configs here
+    if provider["type"] == MetricsProviderType.Uwsgi:
+        return provider.get("use_prometheus", DEFAULT_USE_PROMETHEUS_UWSGI)
+    if provider["type"] == MetricsProviderType.CPU:
+        return provider.get("use_prometheus", DEFAULT_USE_PROMETHEUS_CPU)
 
+    return True
 
-def should_create_gunicorn_scaling_rule(
-    autoscaling_config: AutoscalingParamsDict,
-) -> Tuple[bool, Optional[str]]:
-    """
-    Determines whether we should configure the prometheus adapter for a given service.
-    Returns a 2-tuple of (should_create, reason_to_skip)
-    """
-    if autoscaling_config["metrics_provider"] == "gunicorn":
-        return True, None
-
-    return False, "did not request gunicorn autoscaling"
-
-
-def should_create_piscina_scaling_rule(
-    autoscaling_config: AutoscalingParamsDict,
-) -> Tuple[bool, Optional[str]]:
-    """
-    Determines whether we should configure the prometheus adapter for a given service.
-    Returns a 2-tuple of (should_create, reason_to_skip)
-    """
-    if autoscaling_config["metrics_provider"] == "piscina":
-        return True, None
-    return False, "did not request piscina autoscaling"
 
+def create_instance_scaling_rule(
+    service: str,
+    instance_config: KubernetesDeploymentConfig,
+    metrics_provider_config: MetricsProviderDict,
+    paasta_cluster: str,
+) -> PrometheusAdapterRule:
+    if metrics_provider_config["type"] == MetricsProviderType.CPU:
+        return create_instance_cpu_scaling_rule(
+            service, instance_config, metrics_provider_config, paasta_cluster
+        )
+    if metrics_provider_config["type"] == MetricsProviderType.Uwsgi:
+        return create_instance_uwsgi_scaling_rule(
+            service, instance_config, metrics_provider_config, paasta_cluster
+        )
+    if metrics_provider_config["type"] == MetricsProviderType.Piscina:
+        return create_instance_piscina_scaling_rule(
+            service, instance_config, metrics_provider_config, paasta_cluster
+        )
+    if metrics_provider_config["type"] == MetricsProviderType.Gunicorn:
+        return create_instance_gunicorn_scaling_rule(
+            service, instance_config, metrics_provider_config, paasta_cluster
+        )
+    if metrics_provider_config["type"] == MetricsProviderType.ActiveRequests:
+        return create_instance_active_requests_scaling_rule(
+            service, instance_config, metrics_provider_config, paasta_cluster
+        )
+    if metrics_provider_config["type"] == MetricsProviderType.ArbitraryPromQL:
+        return create_instance_arbitrary_promql_scaling_rule(
+            service, instance_config, metrics_provider_config, paasta_cluster
+        )
```

Comment:
I thought about doing something like this, but I kinda prefer the explicitness of the ifs.... what I _really_ want is a case statement where I can match on the type, so that you never forget to update this in the future, but you need to update to a more recent version of Python for this.

Reply:
i *think* we should be unblocked relatively soon on bumping the minimum python version - but yea :p